### PR TITLE
Add messages.pot update script to pre-commit.

### LINKED ISSUE #8818: Add `messages.pot` updates to pre-commit hook (if possible)
### Describe the problem that you'd like solved

It would be nice if the `messages.pot` file, which stores all the to-be-translated strings and their locations in the current version of the code, could be updated automatically with new PRs to keep everything more up to date for translators. After finishing up a bulk update to the `messages.pot` file for translations (#8817), I was talking with @cdrini about options for automating the process:

### Proposal & Constraints

I'm not sure how the implementation for this would work, or whether it would be possible, but ideally one would bundle in the command that generates a new `messages.pot` file -- `docker compose run --rm -uroot home ./scripts/i18n-messages extract` -- into the already-existing `pre-commit` hook. More info re: updating in the [documentation](https://github.com/internetarchive/openlibrary/wiki/Internationalization-Contributor's-Guide-(i18n)#extracting-strings-from-htmlpython-files-generating-the-pot-file). 

This would be great, as it would mean that any time someone adds a new untranslated string, or modifies a file so the string line numbers are different, the update to the translation file would be included in the PR. Optionally/if possible, it would be great if the translation update could only run if the Python and/or HTML files had been modified, as those are the only changes that would affect it.

Alternatively, either if this isn't possible or just as a stop-gap measure, one could add a note to the Git Cheat Sheet that requests that people run the command themselves as part of the PR if their update has involved adding/changing any on-screen text. If this is the preferred solution or just something that would be good in the meantime, I'd be happy to do that update myself!

### Stakeholders

@cdrini

### PR DESCRIPTION
**Pull request by rebecca-shoptaw**
Closes #8818.

Feature. Automates generation of new `messages.pot` file for translation whenever Python or HTML files that affect `i18n` strings are changed.

### Technical

The implementation is in two parts and mainly confined to two files:
1. `pre-commit-config.yaml` -- Added a new `pre-commit` local hook that runs the `./scripts/i18n-messages extract` script to generate a new `messages.pot` file and is activated by changes to Python or HTML files.

**Note:** This hook has a few dependencies will need to be updated to match the current version when we update `Babel`, `multipart`, and/or `web.py`.

2. `__init__.py` in `openlibrary/i18n` -- Wrote new logic into the message extraction function, using a straightforward `symmetric_difference` compare between the old and new `i18n` messages to make sure that `messages.pot` only gets updated when `i18n` strings have been added/modified/deleted. 

**Note:** This was done to prevent an infinite loop of failure within the pre-commit hook, as the hook will otherwise update the template time field in `messages.pot` on every run, and report a failed run as a result of the file modification, which prevents CI autofixing.

This extra logic also had the nice side effect of ensuring that there will be no `messages.pot` updates for totally unrelated files, as only actual relevant string changes will trigger the re-generation.

### Testing

1. Make a change to a Python or HTML file that adds, removes or modifies `i18n`-formatted text
4. Install `pre-commit` if not installed
5. Commit the change; `pre-commit` should automatically generate a new `messages.pot` file

### Stakeholders

@cdrini

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -91,3 +91,16 @@ repos:
...
+    - id: generate-pot
+      name: Generate POT
+      description: This hook generates the .pot file for internationalization.
+      entry: python ./scripts/i18n-messages extract
+      language: python
+      files: '\.(html|py)$'
+      additional_dependencies:
+        - Babel==2.12.1
+        - multipart==0.2.4
+        - "git+https://github.com/webpy/webpy.git@ed3e92cceb6ed870b224107ea653f48fa7fc2d0a#egg=web-py"
```
Comment (scottbarnes) [inline .pre-commit-config.yaml:106]:
```suggestion
        - "git+https://github.com/webpy/webpy.git@d3649322b85777b291ac2b7b3699fb6fc839e382#egg=web-py"
```

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -91,3 +91,16 @@ repos:
...
       args: [--fix]
+
+  - repo: local
+    hooks:
+    - id: generate-pot
+      name: Generate POT
+      description: This hook generates the .pot file for internationalization.
+      entry: python ./scripts/i18n-messages extract
+      language: python
+      files: '\.(html|py)$'
```
Comment (scottbarnes) [inline .pre-commit-config.yaml:102]:
```suggestion
pass_filenames: false
```
This should stop `pre-commit` from running the command multiple times.

**Review by scottbarnes — COMMENTED** [review]
I think the `SyntaxWarning: invalid escape sequence '\ '` is coming from something that `i18n-messages extract` is calling, rather than from `pre-commit`.

I don't think this is the solution, but with the change to use `pass_filenames: false` instead of `files`, and the following, `pre-commit` shows `Passed` for `Generate POT`:

```diff
diff --git a/scripts/i18n-messages b/scripts/i18n-messages
index 98f28e5d6..4085c927c 100755
--- a/scripts/i18n-messages
+++ b/scripts/i18n-messages
@@ -4,6 +4,10 @@ macros and write to openlibrary/i18n/messages.pot file.
 """
 import _init_path  # noqa: F401  Imported for its side effect of setting PYTHONPATH
 
+import warnings
+
+warnings.simplefilter('ignore')
+
 import sys
 from openlibrary import i18n
 ```

Perhaps somewhere inside is a regex that needs `r` in front of the string. Not sure.

**Comment by rebecca-shoptaw** [comment]
pre-commit.ci autofix

**Review by cdrini — CHANGES_REQUESTED** [review]
Awesome, thank you @rebecca-shoptaw !! I tested this in gitpod/locally, and:

- ✅ When no html/python files staged it doesn't run
- ✅ When HTML file staged with no i18n lines changed, it runs but doesn't save
- ✅ When HTML file staged with i18n lines moved but not modified, it runs but doesn't save
- ✅ When HTML file staged with i18n lines modified, it runs and saves

There's one case I realized would cause some issues; on my machine I have a bunch of untracked files, some of which are html/python! These are getting added to the messages.pot file when I do a commit, which is not what we want. I reckon we should try to exclude untracked/unstaged files when running from pre-commit. Maybe a `--pre-commit` option that adds these exclusions?

```py
    # Exclude untracked git files.
    # git ls-files --others --exclude-standard
    untracked_files = {
        Path(line)
        for dir in dirs
        for line in subprocess.run(
            ['git', 'ls-files', '--others', '--exclude-standard', dir],
            stdout=subprocess.PIPE,
            text=True,
        ).stdout.split('\n')
        if line.endswith(('.py', '.html'))
    }

    # Also exclude unstaged git files.
    # git diff --name-only
    unstaged_files = {
        Path(line)
        for line in subprocess.run(
            ['git', 'diff', '--name-only'],
            stdout=subprocess.PIPE,
            text=True,
        ).stdout.split('\n')
        if line.endswith(('.py', '.html'))
    }

    excluded_files = untracked_files | unstaged_files
```

And then when it builds the catalog makes sure to toss these out?

Otherwise a few small code cleanups/fixes!

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -91,3 +91,17 @@ repos:
...
+    hooks:
+    - id: generate-pot
+      name: Generate POT
+      description: This hook generates the .pot file for internationalization.
+      entry: python ./scripts/i18n-messages extract
+      language: python
+      types_or: [python, html]
+      pass_filenames: false
+      additional_dependencies:
+        - Babel==2.12.1
```
[the remark this answers is not in the document]
Comment (cdrini) [reply .pre-commit-config.yaml:105]: Actually nevermind this is an issue with all the dependencies in this file, sigh!

**Review by cdrini — APPROVED** [review]
Lgtm! One small change

**Comment by cdrini** [comment]
We talked Creation-Date potentially causing a few headaches via merge conflicts ; we decided to create a separate issue to fix the Creation-Date to the first day of the current year; this should hopefully not interfere with any translator flows, whilst reducing conflict headaches for us!

**Pull request merged**